### ISSUE TITLE
Issue 33171: The Golang Ternary Schism

### ISSUE THREAD

**Issue by dash-** [golang/go#46430]
This is a duplicate of the following issue:  [https://github.com/dash-/go/issues/1](https://github.com/dash-/go/issues/1)

---

I have temporarily forked Golang/Go and created this issue because the Golang overlords have stopped playing nice and they can no longer be trusted to host this conversation.

Regarding [Go Issue #33171](https://github.com/golang/go/issues/33171), I propose again there is a third option: compromise.

On July 18, 2019, Github user `phrounz` stood up against the "Go convention and language's designers arguments” in the FAQ regarding the ternary conditional operator ("ternary operator"). The user stated specific examples where the ternary operator could improve readability. As of today, May 27, 2021 (the date the issue was locked as resolved and conversation limited), there were 105 thumbs up to this, and 107 thumbs against, with 5 frowny faces.  I personally suspect Google employees filled out the last of the 107 thumbs up and stopped after passing the 105 thumbs down, but that suspicion is admittedly groundless. (I'm sure I could investigate, it is accessible information, but I'm too lazy and just don't really care.)

Rather than hear reason or find compromise, the "language's designers" and their defenders ("the Golang overlords"/"the majority party") planted themselves stubbornly against `phrounz` and the oncoming horde of software engineers forced into Golang projects by their corporate circumstances whose practical concern is just occasional convenience of the ternary operator in what is otherwise an "I could take it or leave it" programming language ("the minority party").

Having seen 8 months of this debate upon my arrival to the issue on March 6, 2020, I offered a **compromise**, a **prediction**, and a **warning**. My words were met by defensiveness, ego, stubbornness, ignorance, and delusion.

The **compromise** I suggested was allowing the ternary operator to be enabled via configuration, even if it is off by default. That would have relieved the massive social tension that was building, and I believe it would have resolved the issue sufficiently for the minority party without totally sacrificing the values of the majority party, and thus manage to dispense the social energy without a social explosion. That is how a technology leader that understands people and how to resolve these kinds of issues masterfully handles these very delicate situations. The default of "off" is your win, which should be good enough for you. If you are truly right, it will remain off in the wild on average. If you are wrong, it will not remain off on average, because the world will decide you are wrong. I suspect, however, the world will just schism, and there will be about a fifty percent chance the option will be turned off no matter where you go (assuming the 105 thumbs up vs 107 thumbs down was legitimate and natural). These types of minor schisms happen all the time in software for precisely this reason, particularly in the realm of linting. You want that schism to occur out there in the wild, away from the language. Otherwise, the language will schism, and you will be forked, which was my **prediction**.

This brings me to the **warning**. I said, “It is a mistake to think that a language should be the product of a small number of architects that know better without a deep connection to those that are actually using the language.” I was surprised this was the only thing griesemer responded to, essentially saying he felt like I was wrong because the overlords talk amongst themselves and the majority party (my apologies if I’m mis-paraphrasing here). To which I would note that no, I mean really hearing all the signals coming in. By the time I arrived, the minority party was effectively muted. Starting a reply with “the conversation is already over” is not okay. You actually need to keep hearing our feedback, every one of us. We are that signal telling you something is wrong. Choosing to mute us rather than solve the underlying issue is like covering up screams with music, especially when the screams are not yours but you are the one causing the screams, and the one that can resolve the matter. In fact, you should be working to find the screams because it takes tons of effort for us to get the screams close to you. Most of the screams are so far away you’ve no hope of hearing them unless you really look. Try googling “hate golang” or “golang ternary hate” and really read the first forty or so results and really understand the loves and the frustrations and pain, and then realize that some of those people don’t want to have to use your language but they are basically forced into it because of their jobs that switched to Golang from something else and they got a “learn this or learn about your next employer” ultimatum. (This wasn’t me, btw. I learn new programming languages like other folks learn new songs. But I’ve read a lot of people’s feedback about Golang while investigating this issue.)

So yes, you should want to know about all of these people’s feedback, especially when it’s them screaming in paper cut agony. Choosing not to hear it because you disagree with it is a bad choice. It lacks empathy. Not only is empathy moral, but it actually helps us constantly make the smart choice by keeping us on the same side as society. Ignoring the minority party is going to hurt the minority party in the short term, but in the long term, the overlords are going to end up looking like the bad guy, and that’s going to hurt the overlords. **You can either choose to not hurt others because of empathy, or because you recognize that your lack of empathy will eventually lead others to conspire against you to your ruin, but either way, you should choose to not hurt others.** Which is actually a really nice way of restating my **warning**, even though it might look very different now that I have spent so many words describing it.

(And I feel like this warning applies so widely to Google vs Society in our modern times, may this serve as a warning from the microcosmic to the macrocosmic as well. Take heed, Google.)

Metaphorically, Golang is a lake built and run by Google, which is filled with paper-cut razors and signs welcoming us into the water. Enough people go that many others are swept along with them, and issue #33171 was serving as the complaint department where people would go to complain about the razors. And they continued to complain. And that took energy to deal with. And the overlords don’t love their energy being sucked into this issue, particularly because they incorrectly see it as a non-issue.

So today, May 27, 2021, after `vitalif` joined the conversation and delivered the final straw by pointing out yet another case where the ternary operator improved readability, Golang closed the complaint department by closing and locking issue #33171 as resolved. It is not.

Intentionally repeating myself here: Ignoring the minority party is going to hurt the folks screaming in the short term, but in the long term, you’re very much going to end up looking like the bad guy, and that’s going to hurt you. And Golang. And I very much feel justified saying this is incompetence by the Golang overlords.

Here’s why: There really is no right choice between the ternary conditional operator or not. If you google it or scour the scholarly databases of academia, there is certainly not statistically significant evidence suggesting a superior position within this schism. Hence, the reason for the existence of the schism. There’s a whole lot of opinions, and they vary widely and deeply. It’s like Vim vs Emacs all over again, except there really is no superior position here. (Vim is obviously superior, lol.)

What is evidence-based is the economics of the matter. The lack of a proper right or wrong naturally produces developers randomly on both sides of a schism (think statistical mechanics). That schism has existed since the ternary conditional operator was first introduced. But many languages (and most popular languages) now allow it, which is basically not taking a position in the matter. Those languages don’t require it’s use or disallow its alternatives. And individual projects can easily disallow its use if those in charge don’t want it. So despite the way it looks, most popular languages are actually quite agnostic about the ternary conditional operator. Golang made a different choice, I suspect because of a combination of ignorance and ego. Upon its conception, Golang decided to take a position in the already existing schism (despite no real evidence of a superior side), thereby coupling the language to the schism. It is a foregone conclusion that either Golang will defeat/survive the schism, or the schism will rip it in half (schisming it). It’s obvious to me that the overlords will lose here. They are up against at least anywhere near half of all their future users. They just don’t have enough social power to win that fight. It’s not clear to me that any organization of humans on earth is that powerful. Some kind of super dictator power might be sufficient, but software developers are practically wizards or lower deities in our modern society. I wouldn’t bet against half of them in literally any fight, except maybe against the other half, but in this case I promise the other half does not care enough.

No, if Google does not change its position and improve its behavior, market forces (demand) will eventually win.

My prediction in March 2020 came from my knowledge of economics, not computer science -- but it's just easy to see the incredible forces of demand on the one side, and the quite impressive stubbornness of the Golang overlords on the other. It doesn't take much to know that something is going to buckle with fairly significant energy.  But there is always a spark that causes the cascade. I am now predicting that today's event was that spark. Of course there is a chance to save it and prevent the fork by allowing the conversation to continue in a more fair environment outside of the Golang overlords' control. Perhaps the overlords won't even show up, in which case I do predict the fork is inevitable. But, perhaps an accord can be reached. I’m hopeful.

Finally, I now predict that if a compromise is not reached between the parties, the language will be forked and the new language will be called “De” (pronounced “Day”).  This is a play on words, similar to how Go is the first two letters of Google, but because many people are increasingly looking at Google like a Demon or Devil, the letters would be “De” instead. So it doubles as a diss against Google for behaving the way it is behaving.

**Comment by ALTree** [comment]
Nice blog post!

> I personally suspect Google employees filled out the last of the 107 thumbs up and stopped after passing the 105 thumbs down, but that suspicion is admittedly groundless. (I'm sure I could investigate, it is accessible information, but I'm too lazy and just don't really care.)

This seems unlikely... but if you ever find time to investigate the matter, please report back on golang-dev. I would personally love to hear more about this.

> The compromise I suggested was allowing the ternary operator to be enabled via configuration, even if it is off by default.

Go is a fairly opinionated language. Essentially every single proposal to significantly alter the language by making it "configurable" (thus introducing "language dialects") has been turned down. Since your post here doesn't add much to the discussion at #33171, I'm closing here as a dup of #33171.

**Comment by ianlancetaylor** [comment]
I'll just add that the way to change a proposal decision is to provide new information.  I don't see any new information here.  Continuing discussion without new information just takes time without making progress.  We have to be able to make a decision and move on, without continuously discussing the same issues.  Thanks, and thanks for taking the time to write this.

**Comment by dash-** [comment]
@ianlancetaylor @ALTree 

Thanks for reading and the reply. Sorry I'm so verbose. I assumed it was preferred, but perhaps not. If not, perhaps consider making Go less verbose by adding ternary conditional operators.

There's no more examples I need to give about where a ternary conditional is better than an if statement. Plenty of people made those examples already, and had someone else say "oh I prefer this really verbose and terrible code, so no ternary." Because this is all just opinions. There isn't a right or wrong there, and that's why I'm not arguing about that part.

My goal of the prior summary was not to provide new information, but rather to sum up the previous conversation for the new venue outside overlord control. I did add more detail to my previous points, added that the conversation is actually taking place elsewhere now because the Golang overlords have been misbehaving and cannot be trusted to host the conversation, and added a prediction about the name of the fork. Although I don't have a lot of faith in that name prediction. Those kinds of predictions are really tough.

That having been said, I did introduce some new thoughts around this issue's relation to human suffering. It's not the kind of information you're looking for, but that's my point. You're not seeing it because you're in the wrong paradigm. I've been a developer for more than a couple decades, but it's my people experience as a lead, principal, and director that I'm speaking from. This conversation is about people, and is not so much about computer science. Otherwise, you're putting computer science before people, and that's fundamentally wrong (both morally and strategically).

People in the general software development community are being subjected against their will to your terrible opinions. Subjected by their jobs and their conditions, their bosses and their teachers. It causes them undue stress, increasing their cortisol, shortening their lifespan. Some of them quit their jobs. Others are fired as a result. You should read **their** blog entries. I even suspect it may have played a factor in at least one suicide. I don't have stats here, but even a conservative estimate would reveal that Golang has negatively affected at least several thousand of its ~1.1M developers. You're doing that to them on purpose, and acting like they are just overreacting. I take it you're not utilitarians. If you can't empathize with them, you should seek therapy. It is dangerous to not feel empathy when you hear about other people experiencing pain, and it should disqualify you from making product decisions. Even product decisions for a programming language.

Your terrible opinions are terrible because they are not evidence-based, and so when half of the world randomly disagrees because there is no right or wrong, you've made it feel like sandpaper on the skin to use your product. You could have been agnostic about the ternary conditional operator like the other popular languages and just included it outright. Instead you're torturing human beings.

Occasionally one of them finds the complaint department, and so you've closed the complaint department -- because when you're torturing people, having a complaint department seems like a wasteful expense.  It's not, it's actually how you suppress the insurrection, and when you close the complaint department that's when the insurrection really heats up. Something is going to happen. This isn't my first rodeo here, and I'm speaking from experience, but I don't know for sure what is going to happen. I predicted a fork, but it doesn't have to be a fork.

Pain results in reaction, and frequently overreaction. There will be an effect of this. Your attempts to "win" or "quell" all the little nitpicking fights of the software world are admirable, but you don't understand all the sciences involved well enough to actually win the very very difficult game you're playing. You're just using computer science, and it's not the dominant science here. Social sciences dominate here.

Human emotion is incredibly powerful when synchronized, so it's easy to predict the energy will go somewhere. It's far harder to predict where it will go. It might not be a fork. It could be a movement, but probably not. It could be violence, but probably not. It could be legislation, but probably not. It could just cause a little annoyance until Golang becomes less popular as new, better, evidence-based languages pop up, but probably not. Maybe it will just result in a preprocessor. That actually seems equally likely as a fork. Maybe even more likely.

One of the ideas I keep circling around and not saying is it's not up to the Golang overlords. It's up to us, the Golang user community. Y'all acting otherwise is ignorance. Me saying this is not what makes it so. It being so is what makes it so. The Go overlords will comply with the community eventually, because the community is everything. What's more, the community has powers beyond the overlords' control. We can fork, but we can also do worse.

It is this last bit that is what this new conversation is about. I'm basically over discussing the ternary operator. I'm mostly trying to negotiate a compromise on behalf of the minority party, but part of that is also having the discussion about what we, as the minority party, will do if compromise cannot be made.

**Comment by dash-** [comment]
Oh yeah, I guess I never really mentioned why I ever came to the original issue. I was relatively new in a position working as a Director at a medium sized corporation at the time, and the CTO above me insisted we use Go for new backend work instead of .NET, which the CTO and I both hated. I had enjoyed my dalliances with Go so the order was unopposed by me. I pushed it onto my subordinates, who pushed it onto theirs, and we started to dev backend in Go rather than .NET. A significant improvement, I thought. "These fortunate devs under me who get to finally use a modern tech stack and not just go down the dead-end of .NET development," I thought, at the time. But then I got pushback after awhile about these kinds of issues, things like unnecessary verbosity, poor readability, etc, and was blown away. That's when I came here back in March, 2020. Eventually, we realized that it wasn't going to change on any reasonable timescale, and we eventually just started looking at different options.  The CTO left and then I got cancer and left.  Last I heard they were back on .NET. Fun story huh.
